File: src/BreakfastProvider.Api/Reporting/EventGridWebhookController.cs
using System.Text.Json;
using Microsoft.AspNetCore.Mvc;

namespace BreakfastProvider.Api.Reporting;

/// <summary>
/// Receives EventGrid events delivered via webhook subscription.
/// In production, Azure EventGrid pushes events to this endpoint.
/// In component tests, events are POSTed directly to simulate delivery.
/// </summary>
[ApiController]
[Route("webhooks/eventgrid")]
public class EventGridWebhookController(
    IReportingIngester ingester,
    ILogger<EventGridWebhookController> logger) : ControllerBase
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNameCaseInsensitive = true
    };

    /// <summary>
    /// Processes an array of EventGrid-schema events. Supports subscription
    /// validation and <c>IngredientDeliveryEvent</c> processing.
    /// </summary>
    [HttpPost]
    [Consumes("application/json")]
    [ProducesResponseType(StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    public async Task<IActionResult> ReceiveEvents([FromBody] JsonElement body, CancellationToken cancellationToken)
    {
        var events = body.ValueKind == JsonValueKind.Array
            ? body.EnumerateArray().ToList()
            : [body];

        foreach (var eventElement in events)
        {
            var eventType = eventElement.TryGetProperty("eventType", out var et)
                ? et.GetString()
                : null;

            if (eventType == "Microsoft.EventGrid.SubscriptionValidationEvent")
            {
                var validationCode = eventElement
                    .GetProperty("data")
                    .GetProperty("validationCode")
                    .GetString();

                return Ok(new { validationResponse = validationCode });
            }

            if (string.Equals(eventType, "IngredientDeliveryEvent", StringComparison.OrdinalIgnoreCase))
            {
                await ProcessIngredientDelivery(eventElement, cancellationToken);
            }
            else
            {
                logger.LogDebug("Ignoring unhandled EventGrid event type: {EventType}", eventType);
            }
        }

        return Ok();
    }

    private async Task ProcessIngredientDelivery(JsonElement eventElement, CancellationToken cancellationToken)
    {
        if (!eventElement.TryGetProperty("data", out var data))
            return;

        var delivery = data.Deserialize<IngredientDeliveryData>(JsonOptions);
        if (delivery is null) return;

        await ingester.IngestIngredientShipmentAsync(
            delivery.DeliveryId,
            delivery.IngredientName,
            delivery.Quantity,
            delivery.DeliveredAt,
            cancellationToken);

        logger.LogInformation("Processed ingredient delivery {DeliveryId} for {Ingredient}",
            delivery.DeliveryId, delivery.IngredientName);
    }

    private class IngredientDeliveryData
    {
        public Guid DeliveryId { get; set; }
        public string IngredientName { get; set; } = string.Empty;
        public decimal Quantity { get; set; }
        public DateTime DeliveredAt { get; set; }
    }
}

File: java/breakfast-provider-app/src/main/java/io/lemonlion/breakfast/reporting/EventGridWebhookController.java
package io.lemonlion.breakfast.reporting;

import com.fasterxml.jackson.databind.JsonNode;
import java.time.Instant;
import java.util.Map;
import java.util.UUID;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

/**
 * Receives EventGrid events delivered via webhook (twin of the C# {@code EventGridWebhookController}).
 * In production Azure EventGrid pushes here; in component tests events are POSTed directly. Supports the
 * subscription-validation handshake and {@code IngredientDeliveryEvent} ingestion.
 */
@RestController
@RequestMapping("/webhooks/eventgrid")
public class EventGridWebhookController {

    private static final Logger log = LoggerFactory.getLogger(EventGridWebhookController.class);

    private final ReportingIngester ingester;

    public EventGridWebhookController(ReportingIngester ingester) {
        this.ingester = ingester;
    }

    @PostMapping(consumes = MediaType.APPLICATION_JSON_VALUE)
    public ResponseEntity<?> receiveEvents(@RequestBody JsonNode body) {
        Iterable<JsonNode> events = body.isArray() ? body : java.util.List.of(body);
        for (JsonNode event : events) {
            String eventType = event.path("eventType").asText(null);

            if ("Microsoft.EventGrid.SubscriptionValidationEvent".equals(eventType)) {
                String validationCode = event.path("data").path("validationCode").asText();
                return ResponseEntity.ok(Map.of("validationResponse", validationCode));
            }

            if (eventType != null && eventType.equalsIgnoreCase("IngredientDeliveryEvent")) {
                processIngredientDelivery(event.path("data"));
            } else {
                log.debug("Ignoring unhandled EventGrid event type: {}", eventType);
            }
        }
        return ResponseEntity.ok().build();
    }

    private void processIngredientDelivery(JsonNode data) {
        if (data.isMissingNode()) {
            return;
        }
        UUID deliveryId = UUID.fromString(data.path("deliveryId").asText());
        String ingredientName = data.path("ingredientName").asText();
        double quantity = data.path("quantity").asDouble();
        Instant deliveredAt = parseInstant(data.path("deliveredAt").asText());
        ingester.ingestIngredientShipment(deliveryId, ingredientName, quantity, deliveredAt);
        log.info("Processed ingredient delivery {} for {}", deliveryId, ingredientName);
    }

    private static Instant parseInstant(String value) {
        if (value == null || value.isBlank()) {
            return Instant.now();
        }
        try {
            return Instant.parse(value);
        } catch (RuntimeException ex) {
            return java.time.OffsetDateTime.parse(value).toInstant();
        }
    }
}
